### PR TITLE
Plafer 1122 fix decorators

### LINKED ISSUE #1122: Decorator bug report
When we execute a decorator within a procedure that only invokes other procedures without executing any other operation such as the example given below, we receive the following error:

```
---- tests::test_prologue::test_get_blk_timestamp stdout ----
thread 'tests::test_prologue::test_get_blk_timestamp' panicked at /Users/f/.cargo/git/checkouts/miden-vm-3db6be6805d1979c/10ed3e3/assembly/src/assembler/span_builder.rs:151:13:
internal error: entered unreachable code: decorators in an empty SPAN block
```

Example:

```
export.account_vault_add_asset
    emit.ADD_ASSET_TO_ACCOUNT_VAULT_EVENT

    # authenticate that the procedure invocation originates from the account context
    exec.authenticate_account_origin
    # => [ASSET]
    # fetch the vault root
    exec.layout::get_acct_vault_root_ptr movdn.4
    # => [ASSET, acct_vault_root_ptr]
    # add the asset to the account vault
    exec.asset_vault::add_asset
    # => [ASSET']
end
```

Adding some additional operations fixes things:

```
export.account_vault_add_asset
    push.1 drop emit.ADD_ASSET_TO_ACCOUNT_VAULT_EVENT

    # authenticate that the procedure invocation originates from the account context
    exec.authenticate_account_origin
    # => [ASSET]
    # fetch the vault root
    exec.layout::get_acct_vault_root_ptr movdn.4
    # => [ASSET, acct_vault_root_ptr]
    # add the asset to the account vault
    exec.asset_vault::add_asset
    # => [ASSET']
end
```

### PR DESCRIPTION
**Pull request by plafer**
Closes #1122

Based on #1473 

Note that I didn't left `DecoratorList` in a similar format as it was (instead of the proposed `DecoratorSpan`) initially to change fewer things in this PR. But then it's not clear to me that we would gain much by using `DecoratorSpan`, since IIUC, it is useful when there are a lot of decorators sit next to each other. Especially after #1457, I don't think this would be very frequent.

### CODE REVIEW

File: assembly/src/assembler/mod.rs
Comment (bobbinth) [inline assembly/src/assembler/mod.rs:747]: If I understood it correctly, the only time we are setting the `after_exit` decorators is at the end of procedure body. I wonder if there is some clever way to get rid of it `after_exit` decorators altogether and append these somewhere else. Not sure if this is possible though.

**Review by bobbinth — COMMENTED** [review]
Looks good! Thank you! I left some comments inline, the main ones are:

- I do think we should move to `DecoratorSpan` model instead of using vectors - but this can be done in a different PR.
- I think there are some serialization inconsistencies. If so, we should try to fix them in this PR.
- I left a couple of suggestion on how we could make the code cleaner - not sure if all of them will work out.

File: assembly/src/library/tests.rs
Comment (plafer) [reply assembly/src/library/tests.rs:163]: As mentioned above, since `bar1` is an `ExternalNode`, and `bar2` is a `JoinNode`, so we will no longer get a single node.

> ideally, we should end up with just 1 node in the forest.

How would we get just 1 node? We will always need to store `bar2`'s implementation, which is 5 nodes. Maybe I'm missing something?

File: assembly/src/assembler/instruction/ext2_ops.rs
Comment (plafer) [reply assembly/src/assembler/instruction/ext2_ops.rs]: Good idea! Implemented.

The minor downside is that it makes `Assembler::compile_body()` slightly less clean, since `basic_block_builder` owns the `MastForestBuilder` mutable reference [here](https://github.com/0xPolygonMiden/miden-vm/blob/cd6fa22a1cffff5e54b77f4b4672bd7816eaed10/assembly/src/assembler/mod.rs#L603), and so we need to access the `MastForestBuilder` through it in the rest of the method. But it makes the rest of the codebase much cleaner, so IMO it's still definitely worth it.

File: core/src/mast/serialization/basic_block_data_builder.rs
Comment (plafer) [reply core/src/mast/serialization/basic_block_data_builder.rs:47]: >  For some decorators this may be fine (as they are smaller than 32 bits), but for things AsmOp (which we currently use for debug info) this would be quite expensive.

In either strategy, we'll need to serialize the `AsmOp` at least once. So we gain by storing each `AsmOp` once, and then encoding the `DecoratorId` in basic blocks only if that specific `AsmOp` is used more than once. But all `AsmOp`s are going to be unique, right?

So IIUC, we end up encoding *more* bytes using your suggestion, since we need, for each `AsmOp`
1. however many bytes for that `AsmOp`
2. an extra 4 bytes for the `DecoratorId`

as opposed to storing the `AsmOp` directly and only storing those bytes (i.e. saving 4 bytes per `AsmOp` used).

Does that sound right?

File: core/src/mast/serialization/basic_block_data_decoder.rs
Comment (plafer) [reply core/src/mast/serialization/basic_block_data_decoder.rs:65]: Ah yes, good catch. But we just need to use `MastForestBuilder` in deserialization instead of `MastForest`, which would fix the problem.

This also made me notice that we are not serializing the decorators of non-basic block nodes... Will fix that too.

File: assembly/src/assembler/mod.rs
Comment (plafer) [reply assembly/src/assembler/mod.rs:222]: split that block into 2 paragraphs

File: core/src/mast/serialization/basic_block_data_decoder.rs
Comment (plafer) [reply core/src/mast/serialization/basic_block_data_decoder.rs:65]: Actually I thought more about it and using `MastForestBuilder` in deserialization might not be a good idea, since using a different version of the builder when deserializing can lead to differences in `MastNodeIds`, and it's probably best to avoid that. We could ensure with a version string that the builder at deserialization time is compatible with the one used at serialization time, but that seems overly complex and bug-prone.

So I think it's worth having larger debug serialized forests using your suggestion (in the next comment) than more compact ones (from the current implementation).

Comment (plafer) [reply core/src/mast/serialization/basic_block_data_decoder.rs:65]: Here is my planned changes that should address this comment (and the other related ones):

- Decorators serialized similar to `MastNode`s
    - i.e. store a `decorators: Vec<DecoratorInfo>`
    - in the forest, `DecoratorId`s point into this buffer
- Change how `BasicBlockNode` is serialized
    - decorators are no longer stored interleaved with operations
    - `MastNodeType::Block` redefined as:
```rs
enum MastNodeType {
    // ...
    Block {
        // Vec<Operation> offset
        offset_ops_offset: u32,
        // offset of `DecoratorList` in `data` buffer, or 0
        offset_decorators_offset: u32
    }
    // ...
}
```

- All non-basic block nodes will not store their before/after decorators directly in `MastNodeInfo`
    - instead, we store these as 2 maps at the forest level (serialized):
    1. `before_enter: BTreeMap<MastNodeId, Vec<DecoratorId>>`
    2. `after_exit: BTreeMap<MastNodeId, Vec<DecoratorId>>`

File: assembly/src/assembler/instruction/ext2_ops.rs
Comment (bobbinth) [reply assembly/src/assembler/instruction/ext2_ops.rs]: Agreed! One nit (feel free to ignore): I noticed that in some cases we've renamed `span` variable into `basic_block_builder` variable. I would probably have shortened it to just `block_builder` or maybe even just `block`.

File: assembly/src/library/tests.rs
Comment (bobbinth) [reply assembly/src/library/tests.rs:164]: Ah - I see! There is a way to solve it - though, not if it makes things more complicated or not. We can define `eq_hash` as a struct:
```rust
pub struct EqHash {
    mast_root: RpoDigest,
    decorator_root: Option<Blake3Digest>,
}
```
`decorator_root` would be computed very similarly to how we compute `eq_hash` now except that if all child nodes have it set to `None`, then the parent node would also have it set to `None`.

This way, both `External` and `Join` node will end up with the same `EqHash` and this would work as before.

File: assembly/src/library/tests.rs
Comment (bobbinth) [reply assembly/src/library/tests.rs:163]: > How would we get just 1 node? We will always need to store `bar2`'s implementation, which is 5 nodes. Maybe I'm missing something?

This is not for now but `bar2`'s implementation does not actually need to be included in the `MastForest` as the roots of `bar1` and `bar2` are the same and re-export should take precedence. But I think this would require implementing a pass over `MastForest` to clean up such occurrences.

File: core/src/mast/serialization/basic_block_data_decoder.rs
Comment (bobbinth) [reply core/src/mast/serialization/basic_block_data_decoder.rs:65]: Makes sense! A couple of questions/comments:

1. What is `DecoratorInfo`? Is it the same as `Decorator`?
2. Why do we say `offset_ops_offset` and `offset_decorators_offset` rather than just `offset_ops_offset` and `decorators_offset`?
3. For the basic block, do we not need to also specify the number of operations and the number of decorators? (i.e., are just offsets enough)?

**Comment by bobbinth** [comment]
One thing we may need to add to this PR (or to a follow-up PR) is related to https://github.com/0xPolygonMiden/miden-vm/issues/1457#issuecomment-2336541497. Basically, when we compute `eq_hash` of a basic block, we should include error codes attached to `assert` operations into it. This is actually causing some real issues in `miden-base` as sometimes error codes don't get reported properly.

File: assembly/src/library/tests.rs
Comment (plafer) [reply assembly/src/library/tests.rs]: Good idea! Implemented.

File: core/src/mast/node/join_node.rs
Comment (plafer) [reply core/src/mast/node/join_node.rs:24]: Created #1490

File: core/src/mast/serialization/basic_block_data_builder.rs
Comment (plafer) [reply core/src/mast/serialization/basic_block_data_builder.rs]: Addressed in #1482

File: core/src/mast/serialization/basic_block_data_decoder.rs
Comment (plafer) [reply core/src/mast/serialization/basic_block_data_decoder.rs]: Addressed in #1482

File: assembly/src/assembler/mod.rs
Comment (plafer) [reply assembly/src/assembler/mod.rs:768]: As discussed before implementing #1482, I don't think we can get rid of `after_exit`

File: assembly/src/assembler/mod.rs
Comment (plafer) [reply assembly/src/assembler/mod.rs]: Done.

Although I was not able to apply the same simplification to `BasicBlockBuilder::try_into_basic_block()`, since the builder is consumed in the process, and so we need to also return the decorators from that call (if any).

File: processor/src/lib.rs
Comment (plafer) [reply processor/src/lib.rs:501]: Created #1491

**Comment by plafer** [comment]
> One thing we may need to add to this PR (or to a follow-up PR) is related to https://github.com/0xPolygonMiden/miden-vm/issues/1457#issuecomment-2336541497. Basically, when we compute eq_hash of a basic block, we should include error codes attached to assert operations into it. This is actually causing some real issues in miden-base as sometimes error codes don't get reported properly.

The `u32` argument to `Assert` and `U32assert2` are now incorporated into the `eq_hash` computation (see [20446b7](https://github.com/0xPolygonMiden/miden-vm/pull/1466/commits/20446b715ebc530f3223c69e9c1cfcb1d24352b0)).

**Comment by bobbinth** [comment]
> The `u32` argument to `Assert` and `U32assert2` are now incorporated into the `eq_hash` computation (see [20446b7](https://github.com/0xPolygonMiden/miden-vm/pull/1466/commits/20446b715ebc530f3223c69e9c1cfcb1d24352b0)).

Nice! There is one other operation which has carries an error code: `MpVerify`. Could we include it as well?

**Comment by plafer** [comment]
> Nice! There is one other operation which has carries an error code: MpVerify. Could we include it as well?

Ah yes - done (in same commit)

File: assembly/src/assembler/mod.rs
Comment (bobbinth) [inline assembly/src/assembler/mod.rs:635]: Not for this PR, but in the future we could move `wrapper` into `ProcedureContext` and then we won't need to pass `None` here and in other similar places.

**Review by bobbinth — APPROVED** [review]
All looks good! Thank you! I left one comment for the future inline.

**Comment by bobbinth** [comment]
btw, as a result of this PR we should also be able to close https://github.com/0xPolygonMiden/miden-vm/issues/832. I think the resolution is that we don't allow procedures with just decorators.

File: assembly/src/assembler/mod.rs
Comment (bitwalker) [inline assembly/src/assembler/mod.rs:725]: This should probably be documented in the Miden documentation, specifically which instructions are decorators, and the fact that they cannot appear in a procedure by themselves.

I'm not sure how likely it is in practice, but I imagine the error would be pretty confusing.

File: assembly/src/errors.rs
Comment (bitwalker) [inline assembly/src/errors.rs:78]: This is an example of an error where using `Report` and ad-hoc diagnostics works better than a concrete error type like `AssemblyError`. This error sets the stage for what is wrong, but is sort of incomplete from the user perspective. Ideally, you'd have the primary label span the procedure body, with a secondary label spanning the first decorator, with a message like `'<opcode>' is a decorator, which cannot appear in a procedure body without at least one non-decorator instruction`, with help text attached to the diagnostic like `try adding a 'nop' instruction at the end of the procedure body to anchor the decorators`.

By using `AssemblyError` for things like this, we make it difficult (if not impossible) to construct these kind of helpful diagnostics, which is why most APIs are designed to return `Report` rather than `AssemblyError`.

Your call on whether you want to make the change, but we should try to get in the habit of thinking about how these errors will be seen and acted on by users, particularly those that don't have as strong a familiarity with Miden Assembly, and may not understand the difference between things like decorators and instructions, since they appear the same way in the source code.

File: core/src/mast/node/basic_block_node/mod.rs
Comment (bitwalker) [inline core/src/mast/node/basic_block_node/mod.rs:299]: FYI `Display` is implemented for `dyn PrettyPrint`, so a reference to any `PrettyPrint` impl automatically implements `Display`. Can still be convenient to impl `Display` explicitly, like done here, but figured it's useful info to have on hand anyway.

File: core/src/mast/node/join_node.rs
Comment (bitwalker) [inline core/src/mast/node/join_node.rs:22]: Looks like most places where we are using `Vec<DecoratorId>` here could use `Box<[DecoratorId]>` instead, and shave off 8 bytes per field (on 64-bit systems). Doesn't look like we ever mutate them in-place, just create/replace. AFAIK `Box::from([]) as Box<[DecoratorId]>` won't actually allocate anything.

Not something we need to change in this PR necessarily, but that'd be the first thing I'd do to reduce the size of `MastNode` when we start trying to trim down the size.

**Review by bitwalker — APPROVED** [review]
Looks good! Left a few notes/suggestions, but nothing that needs to block merging this. Nice work!

File: assembly/src/assembler/mod.rs
Comment (bobbinth) [reply assembly/src/assembler/mod.rs:725]: This can probably go into the [Procedures](https://0xpolygonmiden.github.io/miden-vm/user_docs/assembly/code_organization.html#procedures) section. And we can refer to the a [advice injectors](https://0xpolygonmiden.github.io/miden-vm/user_docs/assembly/io_operations.html#nondeterministic-inputs) as well as list `emit`, `debug`, and `trace` instructions (so that we don't have to list all advice injectors).

File: assembly/src/errors.rs
Comment (bobbinth) [reply assembly/src/errors.rs:78]: I echoed this comment in https://github.com/0xPolygonMiden/miden-vm/issues/1431 so that we can address it when removing `AssemblyError`.

File: assembly/src/assembler/mod.rs
Comment (plafer) [reply assembly/src/assembler/mod.rs:725]: Updated the docs [2e8481e](https://github.com/0xPolygonMiden/miden-vm/pull/1466/commits/2e8481e1b138c21eb31be2fc43021c9277866dc8). With #1457 we can fix it more holistically - the "decorators" name could be removed in favor of "debug instructions" or whatnot, and here in the "debug instructions" section we can state that these do not modify the MAST root, can be stripped away without affecting the behavior of the program, and cannot be the only instructions in a procedure

**Pull request merged**
